### PR TITLE
Feature/refactoring

### PR DESCRIPTION
**Pull request by wtang-8**
**Warning :** Before creating a pull request, make sure that it does not already exists in the [pull requests](../). Thank you.

## Description
Removed the variable "enable_" in all the filters and detectors. The logic to check if the filter is enabled before executing it is now done in the parent class, which is the class Filter.

## Fixes
Link all the related issues from the issue tracker
- Closes: #35 

## How has this been tested ?
Compile the code and see if it generates any errors.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Review by ProgHunter — CHANGES_REQUESTED** [review]
The branch should be merged in develop, not master. 
Otherwise, good refactoring of "enable_". 👍

File: src/proc_image_processing/filters/adaptive_threshold.h [hunk 1]
```diff
@@ -28,8 +27,7 @@ namespace proc_image_processing {
 
     virtual ~AdaptiveThreshold() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/adaptive_threshold.h]: Je changerais pour apply(cv::Mat& image). L'object est un filter, pas besoin de préciser qu'on ApplyFilter, juste apply c'est suffisant. Généralement quand on va instancier notre filter, la variable risque de se nommer filterQuelqueChose, donc quand on fait filterQuelqueChose->apply() ça fait plus de sens que filterQuelqueChose->ApplyFilter(). Aussi pourquoi le first letter est capital ? les fonctions devrait pas starter en minuscules pour réserver ça aux classes?

File: src/proc_image_processing/filters/background_substract.h [hunk 2]
```diff
@@ -32,8 +31,7 @@ namespace proc_image_processing {
 
     virtual ~BackgroundSubstract() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/background_substract.h]: Same

File: src/proc_image_processing/filters/bilateral_filter.h [hunk 3]
```diff
@@ -37,24 +36,15 @@ namespace proc_image_processing {
      *
      * \param image The image to process.
      */
-    void Execute(cv::Mat& image) override {
-      if (enable_()) {
+    void ApplyFilter(cv::Mat& image) override {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/bilateral_filter.h]: Same

File: src/proc_image_processing/filters/blurr.h [hunk 4]
```diff
@@ -33,8 +32,7 @@ namespace proc_image_processing {
 
     virtual ~Blurr() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/blurr.h]: Same

File: src/proc_image_processing/filters/canny.h [hunk 5]
```diff
@@ -27,18 +26,16 @@ namespace proc_image_processing {
 
     virtual ~Canny() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/canny.h]: Same

File: src/proc_image_processing/filters/contrast_brightness.h [hunk 6]
```diff
@@ -32,8 +31,7 @@ namespace proc_image_processing {
 
     virtual ~ContrastBrightness() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/contrast_brightness.h]: Same

File: src/proc_image_processing/filters/convex_hull.h [hunk 7]
```diff
@@ -30,8 +29,7 @@ namespace proc_image_processing {
 
     virtual ~ConvexHull() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/convex_hull.h]: Same

File: src/proc_image_processing/filters/deep_2019.h [hunk 8]
```diff
@@ -38,8 +37,7 @@ namespace proc_image_processing {
 
         virtual ~Deep2019() { image_subscriber_.shutdown(); }
 
-        virtual void Execute(cv::Mat& image) {
-            if (enable_()) {
+        virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/deep_2019.h]: Same

File: src/proc_image_processing/filters/detectors/center_coffin_detector.h [hunk 9]
```diff
@@ -29,8 +28,7 @@ namespace proc_image_processing {
 
         virtual ~CenterCoffinDetector() {}
 
-        virtual void Execute(cv::Mat& image) {
-            if (enable_()) {
+        virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/detectors/center_coffin_detector.h]: Same

File: src/proc_image_processing/filters/detectors/fence_detector.h [hunk 10]
```diff
@@ -43,11 +42,7 @@ namespace proc_image_processing {
 
     virtual ~FenceDetector() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (!enable_()) {
-        return;
-      }
-
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/detectors/fence_detector.h]: Same

File: src/proc_image_processing/filters/detectors/gate_detector.h [hunk 11]
```diff
@@ -68,8 +67,7 @@ namespace proc_image_processing {
 
     virtual ~GateDetector() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/detectors/gate_detector.h]: Same

File: src/proc_image_processing/filters/detectors/handle_detector.h [hunk 12]
```diff
@@ -44,8 +43,7 @@ namespace proc_image_processing {
 
     virtual ~HandleDetector() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/detectors/handle_detector.h]: Same

File: src/proc_image_processing/filters/detectors/pipe_angle_detector.h [hunk 13]
```diff
@@ -33,8 +32,7 @@ namespace proc_image_processing {
 
         virtual ~PipeAngleDetector() {}
 
-        virtual void Execute(cv::Mat& image) {
-            if (enable_()) {
+        virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/detectors/pipe_angle_detector.h]: Same

File: src/proc_image_processing/filters/detectors/vampire_body_detector.h [hunk 14]
```diff
@@ -29,8 +28,7 @@ namespace proc_image_processing {
 
         virtual ~VampireBodyDetector() {}
 
-        virtual void Execute(cv::Mat& image) {
-            if (enable_()) {
+        virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/detectors/vampire_body_detector.h]: Same

File: src/proc_image_processing/filters/detectors/vampire_torpedoes_close_detector.h [hunk 15]
```diff
@@ -33,8 +32,7 @@ namespace proc_image_processing {
 
         virtual ~VampireTorpedoesDetectorClose() {}
 
-        virtual void Execute(cv::Mat& image) {
-            if (enable_()) {
+        virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/detectors/vampire_torpedoes_close_detector.h]: Same

File: src/proc_image_processing/filters/detectors/vampire_torpedoes_detector.h [hunk 16]
```diff
@@ -33,8 +32,7 @@ namespace proc_image_processing {
 
         virtual ~VampireTorpedoesDetector() {}
 
-        virtual void Execute(cv::Mat& image) {
-            if (enable_()) {
+        virtual void ApplyFilter(cv::Mat& image) {
```
Comment (thecanadianroot) [inline src/proc_image_processing/filters/detectors/vampire_torpedoes_detector.h]: Same

**Review by thecanadianroot — CHANGES_REQUESTED** [review]

File: src/proc_image_processing/filters/adaptive_threshold.h (hunk 1 above)
Comment (thecanadianroot) [reply src/proc_image_processing/filters/adaptive_threshold.h]: Aussi, voir code smell de sonarcloud: https://sonarcloud.io/project/issues?branch=feature%2Frefactoring&id=sonia-auv_proc_image_processing&issues=AXpt9H0jxl0da__V49lY&open=AXpt9H0jxl0da__V49lY

**Review by thecanadianroot — CHANGES_REQUESTED** [review]

File: src/proc_image_processing/filters/adaptive_threshold.h (hunk 1 above)
Comment (wtang-8) [reply src/proc_image_processing/filters/adaptive_threshold.h]: Pour le nom des fonctions en PascalCalse vs camelCase, j'ai écrit un message dans Discord pour vérifier avec l'équipement de développement, car les deux façons de faire semblent correctes quoique camelCase est plus populaire.

Comment (wtang-8) [reply src/proc_image_processing/filters/adaptive_threshold.h]: Je viens de voir les updates de Alexandre sur Discord. Les fonctions seront changées en lowerCase.

Comment (wtang-8) [reply src/proc_image_processing/filters/adaptive_threshold.h]: done

File: src/proc_image_processing/filters/background_substract.h (hunk 2 above)
Comment (wtang-8) [reply src/proc_image_processing/filters/background_substract.h]: done

File: src/proc_image_processing/filters/hough_line.h [hunk 17]
```diff
@@ -27,8 +26,7 @@ namespace proc_image_processing {
 
     virtual ~HoughLine() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/hough_line.h]: Should we declare ApplyFilter() in filter's implementation as "final" ?

File: src/proc_image_processing/filters/hsv_threshold.h [hunk 18]
```diff
@@ -39,8 +38,7 @@ namespace proc_image_processing {
 
     virtual ~HSVThreshold() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/hsv_threshold.h]: no virtual needed here

File: src/proc_image_processing/filters/image_accumulator.h [hunk 19]
```diff
@@ -32,8 +31,7 @@ namespace proc_image_processing {
 
     virtual ~ImageAccumulator() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/image_accumulator.h]: no virtual needed here

File: src/proc_image_processing/filters/image_cropper.h [hunk 20]
```diff
@@ -25,19 +24,16 @@ namespace proc_image_processing {
 
     virtual ~ImageCropper() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/image_cropper.h]: no virtual needed here

File: src/proc_image_processing/filters/laplacian.h [hunk 21]
```diff
@@ -26,8 +25,7 @@ namespace proc_image_processing {
 
     virtual ~Laplacian() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/laplacian.h]: no virtual needed here

File: src/proc_image_processing/filters/mission_test_fake_string.h [hunk 22]
```diff
@@ -16,21 +16,17 @@ namespace proc_image_processing {
...
-      enable_("Enable", false, &parameters_),
       _string("String_to_return", "test", &parameters_) {
       SetName("MissionTestFakeString");
     }
 
     virtual ~MissionTestFakeString() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/mission_test_fake_string.h]: no virtual needed here

File: src/proc_image_processing/filters/morphology.h [hunk 23]
```diff
@@ -29,8 +28,7 @@ namespace proc_image_processing {
 
     virtual ~Morphology() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/morphology.h]: no virtual needed here

File: src/proc_image_processing/filters/original_image.h [hunk 24]
```diff
@@ -15,20 +15,15 @@ namespace proc_image_processing {
...
-      : Filter(globalParams), enable_("Enable", false, &parameters_) {
+      : Filter(globalParams) {
       SetName("OriginalImage");
     }
 
     virtual ~OriginalImage() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/original_image.h]: no virtual needed here

File: src/proc_image_processing/filters/remove_mask.h [hunk 25]
```diff
@@ -32,14 +31,11 @@ namespace proc_image_processing {
 
     virtual ~RemoveMask() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/remove_mask.h]: no virtual needed here

File: src/proc_image_processing/filters/rotate.h [hunk 26]
```diff
@@ -25,8 +24,7 @@ namespace proc_image_processing {
 
     virtual ~Rotate() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/rotate.h]: no virtual needed here

File: src/proc_image_processing/filters/schar_adding.h [hunk 27]
```diff
@@ -36,8 +35,7 @@ namespace proc_image_processing {
 
     virtual ~ScharrAdding() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/schar_adding.h]: no virtual needed here

File: src/proc_image_processing/filters/scharr.h [hunk 28]
```diff
@@ -29,8 +28,7 @@ namespace proc_image_processing {
 
     virtual ~Scharr() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/scharr.h]: no virtual needed here

File: src/proc_image_processing/filters/sobel.h [hunk 29]
```diff
@@ -30,8 +29,7 @@ namespace proc_image_processing {
 
     virtual ~Sobel() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/sobel.h]: no virtual needed here

File: src/proc_image_processing/filters/square_detection.h [hunk 30]
```diff
@@ -17,15 +17,15 @@ namespace proc_image_processing {
...
+            : Filter(globalParams),
             N("N", 100, 0, 100, &parameters_),
             thresh("threshold", 100, 0, 100, &parameters_) {
             SetName("SquareDetection");
         }
 
         virtual ~SquareDetection() {}
 
-        virtual void Execute(cv::Mat& image) {
+        virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/square_detection.h]: no virtual needed here

File: src/proc_image_processing/filters/stats_threshold.h [hunk 31]
```diff
@@ -26,8 +25,7 @@ namespace proc_image_processing {
 
     virtual ~StatsThreshold() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/stats_threshold.h]: no virtual needed here

File: src/proc_image_processing/filters/submarine_frame_masker.h [hunk 32]
```diff
@@ -29,8 +28,7 @@ namespace proc_image_processing {
 
     virtual ~SubmarineFrameMasker() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/submarine_frame_masker.h]: no virtual needed here

File: src/proc_image_processing/filters/subtract_all_planes.h [hunk 33]
```diff
@@ -42,8 +41,7 @@ namespace proc_image_processing {
 
     virtual ~SubtractAllPlanes() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/subtract_all_planes.h]: no virtual needed here

File: src/proc_image_processing/filters/subtract_plane_adder.h [hunk 34]
```diff
@@ -45,7 +44,6 @@ namespace proc_image_processing {
     virtual ~SubtractPlaneAdder() {}
 
     virtual void execute(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/subtract_plane_adder.h:47]: replace with "void ApplyFilter(cv::Mat& image) {" ?

File: src/proc_image_processing/filters/test_filter.h [hunk 35]
```diff
@@ -33,19 +32,16 @@ namespace proc_image_processing {
 
     virtual void init() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/test_filter.h]: no virtual needed here

File: src/proc_image_processing/filters/threshold.h [hunk 36]
```diff
@@ -25,8 +24,7 @@ namespace proc_image_processing {
 
     virtual ~Threshold() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/threshold.h]: no virtual needed here

File: src/proc_image_processing/filters/threshold_between.h [hunk 37]
```diff
@@ -26,8 +25,7 @@ namespace proc_image_processing {
 
     virtual ~ThresholdBetween() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/threshold_between.h]: no virtual needed here

File: src/proc_image_processing/filters/white_filter.h [hunk 38]
```diff
@@ -28,7 +28,7 @@ namespace proc_image_processing {
 
         virtual ~WhiteFilter() {}
 
-        virtual void Execute(cv::Mat& image) {
+        virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/white_filter.h]: no virtual needed here

File: src/proc_image_processing/filters/white_filter.h [hunk 39]
```diff
@@ -44,7 +44,6 @@ namespace proc_image_processing {
     private:
         cv::Mat output_image_;
 
-        Parameter<bool> enable_;
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/white_filter.h:46]: The variable "enable_" was not used in this filter. We should be vigilant about its behaviour when adding the condition by calling Execute() on it. Try to verify if it works correctly with this change.

File: src/proc_image_processing/filters/white_noise_takedown.h [hunk 40]
```diff
@@ -30,8 +29,7 @@ namespace proc_image_processing {
 
     virtual ~WhiteNoiseTakedown() {}
 
-    virtual void Execute(cv::Mat& image) {
-      if (enable_()) {
+    virtual void ApplyFilter(cv::Mat& image) {
```
Comment (ProgHunter) [inline src/proc_image_processing/filters/white_noise_takedown.h]: no virtual needed here

**Review by ProgHunter — CHANGES_REQUESTED** [review]
See comments :)

**Comment by ProgHunter** [comment]
Since the PR relies on a too old version of 'develop', we decided to reimplement the changes from the current state of the branch, this will be way more simple to merge.

**Pull request closed without merging**